File: 프로그래머스/python/LV3/등굣길.py
def solution(m, n, puddles):
    answer = 0
    dp = [[0] * (m + 1) for _ in range(n + 1)]
    dp[1][1] = 1

    for i, j in puddles:
        dp[j][i] = -1

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if (i == 1 and j == 1) or dp[i][j] == -1:
                continue
            if dp[i - 1][j] == -1:
                dp[i][j] = dp[i][j - 1]
            elif dp[i][j - 1] == -1:
                dp[i][j] = dp[i - 1][j]
            else:
                dp[i][j] = (dp[i - 1][j] + dp[i][j - 1]) % 1000000007

    return dp[n][m]

File: 프로그래머스/javascript/LV3/등굣길.js
function solution(m, n, puddles) {
  const board = Array.from({ length: n }, () => Array(m).fill(0));
  const dp = Array.from({ length: n }, () => Array(m).fill(0));
  dp[0][0] = 1;
  for (const [x, y] of puddles) {
    board[y - 1][x - 1] = 1;
  }

  // 각 단계에서 나머지를 구해야 효율성 테스트 통과
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (i === 0 && j === 0) continue;
      if (board[i][j] === 1) continue;
      if (i === 0) {
        dp[i][j] = dp[i][j - 1] % 1000000007;
      } else if (j === 0) {
        dp[i][j] = dp[i - 1][j] % 1000000007;
      } else {
        dp[i][j] = (dp[i - 1][j] + dp[i][j - 1]) % 1000000007;
      }
    }
  }
  return dp[n - 1][m - 1];
}
